canonical/cloud-init
# Correct documentation and testcase data for some user-data YAML

## Proposed Commit Message

For cc_users_groups the user setting "expiredate" **must be quoted** in order for the
relevant flag and value to be then passed to the useradd command. It its vaiue is not
quoted then it is treated as Python type datetime.date and in ```cloudinit/distros/__init__.py```
the below "_is it a string_" condition fails and so no "--expiredate" parameter is passed
to useradd and therefore it has no effect:

```
if key in useradd_opts and val and isinstance(val, str):
    useradd_cmd.extend([useradd_opts[key], val])
```

For cc_users_groups, the user setting "inactive" does not actually disable
accounts, the useradd "--inactive" option actually defines the number of
days after password expiry that users can still login. So I have changed
the docs to show it taking a **quoted** value of days (which works with the
current code) rather than a boolean value. **The quotes are necessary**,
like expiredate above, so that the value is also passed to the useradd command.

For cc_power_state_change.py the "delay" setting value needs to have quotes around
it **as otherwise its leading plus sign will be stripped off**.

## Additional Context

freebsd.py, which doesn't use useradd, *already* has broken handling of
"inactive" as it passes a non-existant "-E" option to the "pw useradd"
command - it should actually pass "-e" instead followed by ```+<DAYS>d```.

I talked to @goneri and @OddBloke about this on IRC.

The existing testcases for users haven't picked up on "expiredate" and
"inactive" not working as they don't check/mock /etc/shadow where
these values end up. I see the testcases using "getent passwd" but
**"getent shadow" will not work when running testcases as a non-root user**.

## Test Steps

User-data:

```
power_state:
  delay: '+5'
users:
  - name: testuser
    expiredate: '2020-12-25'
    gecos: Test User
    inactive: '5'
```

Afterwards check the "testuser" entry in /etc/shadow:

```
grep testuser /etc/shadow
testuser:<ENCRYPTED PASSWORD>:<LAST PWD CHANGE>:<MIN AGE>:<MAX AGE>:<WARN>:<INACTIVITY>:<EXPIRE DATE>:
```
where ```<INACTIVITY>``` is the number specified by "inactive" user-data and ```<EXPIRE DATE>``` is the value of "expiredate" as number of days from 1/1/1970.

## Checklist:
